src/rust/inetstack/protocols/udp/peer.rs

```diff
@@ -166,63 +160,98 @@ impl<const N: usize> UdpPeer<N> {
 
     /// Asynchronously send unsent datagrams to remote peer.
     async fn background_sender(
-        rt: Rc<dyn NetworkRuntime<N>>,
         local_ipv4_addr: Ipv4Addr,
         local_link_addr: MacAddress,
         offload_checksum: bool,
         arp: ArpPeer<N>,
-        mut rx: SharedQueue<SharedQueueSlot<DemiBuffer>>,
+        mut rx: SharedQueue<N>,
     ) {
         loop {
             // Grab next unsent datagram.
             match rx.pop().await {
                 // Resolve remote address.
-                Ok(SharedQueueSlot { local, remote, data }) => match arp.query(remote.ip().clone()).await {
-                    // Send datagram.
-                    Ok(link_addr) => {
-                        Self::do_send(
-                            rt.clone(),
-                            local_ipv4_addr,
-                            local_link_addr,
-                            link_addr,
-                            data,
-                            &local,
-                            &remote,
-                            offload_checksum,
-                        );
-                    },
-                    // ARP query failed.
-                    Err(e) => warn!("Failed to send UDP datagram: {:?}", e),
+                Ok(slot) => {
+                    match arp.query(slot.get_remote().ip().clone()).await {
+                        Ok(link_addr) => {
+                            if let Err(e) = slot.get_queue().pushto(
+                                local_ipv4_addr,
+                                &slot.get_remote(),
+                                local_link_addr,
+                                link_addr,
+                                slot.get_data(),
```

Comment:
Amy we rename this to `rmeote_link_addr`?

Reply:
yup got it